craft_application/commands/remote.py

```diff
@@ -99,11 +103,27 @@ def _run(
             permanent=True,
         )
 
-        if not parsed_args.launchpad_accept_public_upload and not emit.confirm(
-            _CONFIRMATION_PROMPT, default=False
+        if (
+            not parsed_args.launchpad_accept_public_upload
+            and (
+                not parsed_args.project
+                or not self._services.remote_build.is_project_private()
+            )
+            and not emit.confirm(_CONFIRMATION_PROMPT, default=False)
         ):
-            emit.message("Cannot proceed without accepting a public upload.")
-            return 77  # permission denied from sysexits.h
+            raise errors.RemoteBuildError(
+                "Remote build needs explicit acknowledgement that data sent to build servers "
+                "is public.",
+                details=(
+                    "In non-interactive runs, please use the option "
+                    "`--launchpad-accept-public-upload`."
+                ),
+                reportable=False,
+                retcode=77,
```

Comment:
```suggestion
                retcode=77,  # os.EX_NOPERM
```

For when we can change this :-)